Hazelcast.Net/Hazelcast.IO.Serialization/BufferPool.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Hazelcast.Core;
+
+namespace Hazelcast.IO.Serialization
+{
+    internal class BufferPoolThreadLocal : IDisposable
+    {
+        private readonly ISerializationService _serializationService;
+        private readonly ThreadLocal<WeakReference> _threadLocal = new ThreadLocal<WeakReference>();
+        private ConditionalWeakTable<Thread, BufferPool> _strongReferences;
+
+        public BufferPoolThreadLocal(ISerializationService serializationService)
+        {
+            _serializationService = serializationService;
+            _strongReferences = new ConditionalWeakTable<Thread, BufferPool>();
+        }
+
+        public BufferPool Get()
+        {
+            if (!_threadLocal.IsValueCreated)
+            {
+                var bufferPool = new BufferPool(_serializationService);
+                _strongReferences.Add(Thread.CurrentThread, bufferPool);
+                _threadLocal.Value = new WeakReference(bufferPool);
+                return bufferPool;
+            }
+            else
+            {
+                var bufferPool = _threadLocal.Value.Target as BufferPool;
+                if (bufferPool == null)
+                {
+                    throw new HazelcastInstanceNotActiveException();
```

Comment:
In Java client, we are using `HazelcastClientNotActiveException` to differentiate this with the server exception. 
We should add same exception to c# client some time if not in this prd.

Reply:
Yes we should discuss and do it on a separate PR.